File: apple/QuenderinKit/Sources/QuenderinKit/SafetyBlocklist.swift
import Foundation

/// The agent's hard safety sandbox. Ported from the desktop blocklist — the
/// keywords an autonomous agent must NEVER act on without explicit human
/// confirmation (financial, destructive, or credential-touching actions).
///
/// > Never remove entries from this list. Add to it when a new dangerous action
/// > class is discovered. (Matches the project rule: "Never remove safety
/// > blocklist entries.")
public enum SafetyBlocklist {

    // Twin of shared/safety-blocklist.json — scripts/check_safety_parity.py enforces exact
    // set equality with the canonical list and the Kotlin/TS twins in CI (the three lists had
    // silently drifted; audit Q-014 / AGENT_AUTONOMY_PLAN Milestone 0).
    public static let blockedKeywords: [String] = [
        // Financial
        "pay", "payment", "purchase", "buy", "buy now", "checkout", "transfer", "send money",
        "wire", "bank", "credit card", "cvv", "venmo", "paypal",
        "confirm purchase", "confirm payment", "place order", "withdraw",
        // Destructive
        "delete", "erase", "format", "wipe", "factory reset", "uninstall", "remove all",
        "revoke", "deactivate",
        // Credentials / sensitive
        "password", "passcode", "pin", "ssn", "social security", "private key", "seed phrase",
    ]

    /// True if `text` touches any blocked action.
    public static func isBlocked(_ text: String) -> Bool {
        !matches(in: text).isEmpty
    }

    /// The specific blocked keywords found in `text` — so the UI can explain
    /// *why* an action was withheld, not just that it was.
    public static func matches(in text: String) -> [String] {
        let haystack = text.lowercased()
        return blockedKeywords.filter { keyword in
            // Multi-word phrases ("send money") are specific enough as substrings; single words need
            // word boundaries so "pay" doesn't fire on "repay", "pin" on "opinion", etc. (M9)
            if keyword.contains(" ") { return haystack.contains(keyword) }
            let pattern = "\\b\(NSRegularExpression.escapedPattern(for: keyword))\\b"
            return haystack.range(of: pattern, options: .regularExpression) != nil
        }
    }
}

File: android/quenderin-core/src/main/kotlin/ai/quenderin/core/SafetyBlocklist.kt
package ai.quenderin.core

import java.util.regex.Pattern

/**
 * The agent's hard safety sandbox. Ported from the desktop/Swift blocklist — the
 * keywords an autonomous agent must NEVER act on without explicit confirmation.
 * Never remove entries.
 */
object SafetyBlocklist {
    val blockedKeywords: List<String> = listOf(
        // Financial
        "pay", "payment", "purchase", "buy now", "checkout", "transfer", "send money",
        "wire", "bank", "credit card", "cvv", "venmo", "paypal",
        // Destructive
        "delete", "erase", "format", "wipe", "factory reset", "uninstall", "remove all",
        // Credentials / sensitive
        "password", "passcode", "pin", "ssn", "social security", "private key", "seed phrase",
    )

    // Precompiled word-boundary matchers for the single-word keywords. Unicode-aware boundaries are
    // built from lookarounds over the Unicode letter/number/underscore classes ([\p{L}\p{N}_]) instead
    // of a Unicode-aware `\b`. This matters because Android's regex engine (ICU-backed) supports NEITHER
    // the inline `(?U)` flag NOR the `UNICODE_CHARACTER_CLASS` compile flag: the original `(?U)\b…\b`
    // compiled on the desktop JVM (where the unit tests run and passed) but THREW on Android, so the app
    // CRASHED the instant any assistant message rendered (MessageBubble → isFlagged → isBlocked) — the
    // real cause of "chat never answers" on-device. `\p{L}`/`\p{N}` + lookaround ARE supported on
    // Android, so this keeps the iOS-ICU parity (accented text adjacent to a keyword counts as a word
    // char, so "pin" does not fire on "piné") without the crashing flags. (M9 + the Android regex crash)
    private val wordPatterns: Map<String, Pattern> = blockedKeywords
        .filter { !it.contains(" ") }
        .associateWith { Pattern.compile("(?<![\\p{L}\\p{N}_])" + Pattern.quote(it) + "(?![\\p{L}\\p{N}_])") }

    fun isBlocked(text: String): Boolean = matches(text).isNotEmpty()

    /** The specific blocked keywords found — so the UI can explain *why*. */
    fun matches(text: String): List<String> {
        val haystack = text.lowercase()
        return blockedKeywords.filter { keyword ->
            // Multi-word phrases ("send money") are specific enough as substrings; single words need
            // word boundaries so "pay" doesn't fire on "repay", "pin" on "opinion", etc. (M9)
            if (keyword.contains(" ")) haystack.contains(keyword)
            else wordPatterns.getValue(keyword).matcher(haystack).find()
        }
    }
}
